tests/e2e/action_rapid_test.go

```diff
@@ -0,0 +1,501 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/stretchr/testify/require"
+	"pgregory.net/rapid"
+)
+
+// This file contains tests for serialization/deserialization of actions.
+// The tests are written using the rapid testing library, which allows us to
+// generate arbitrary actions and test that they can be serialized and
+// deserialized without error.
+// The generators for the various actions are defined in this file, and
+// essentially tell rapid how to build these actions.
+
+func TestActionMarshalling(t *testing.T) {
+	rapid.Check(t, func(t *rapid.T) {
+		action := GetActionGen().Draw(t, "Action")
+		err := MarshalAndUnmarshalAction(action)
+		if err != nil {
+			t.Fatalf("error marshalling and unmarshalling action: %v", err)
+		}
+	})
+}
+
+func MarshalAndUnmarshalAction(action interface{}) error {
+	// wraps the action with a step, since it needs custom unmarshalling that is called by the step unmarshaller
+	step := Step{
+		Action: action,
+	}
+	jsonobj, err := json.Marshal(step)
+	if err != nil {
+		return fmt.Errorf("error marshalling action inside step: %v", err)
+	}
+
+	var got Step
+	err = json.Unmarshal(jsonobj, &got)
+	if err != nil {
+		return fmt.Errorf("error unmarshalling action inside step: %v", err)
+	}
+
+	diff := cmp.Diff(step, got)
+	if diff != "" {
+		return fmt.Errorf("got (-), want (+): %v", diff)
+	}
+
+	return nil
+}
+
+// This needs to be adjusted manually when new actions are added and should
+// include generators for all actions that are mentioned in main.go/runStep.
+func GetActionGen() *rapid.Generator[any] {
+	return rapid.OneOf(
+		GetStartSovereignChainActionGen().AsAny(),
+		GetSubmitLegacyUpgradeProposalActionGen().AsAny(),
+		GetWaitUntilBlockActionGen().AsAny(),
+		GetChangeoverChainActionGen().AsAny(),
+		GetSendTokensActionGen().AsAny(),
+		GetStartChainActionGen().AsAny(),
+		GetSubmitTextProposalActionGen().AsAny(),
+		GetSubmitConsumerAdditionProposalActionGen().AsAny(),
+		GetSubmitConsumerRemovalProposalActionGen().AsAny(),
+		GetSubmitParamChangeProposalActionGen().AsAny(),
+		GetSubmitEquivocationProposalActionGen().AsAny(),
+		GetVoteGovProposalActionGen().AsAny(),
+		GetStartConsumerChainActionGen().AsAny(),
+		GetAddChainToRelayerActionGen().AsAny(),
+		GetAddIbcConnectionActionGen().AsAny(),
+		GetAddIbcChannelActionGen().AsAny(),
+		GetStartRelayerActionGen().AsAny(),
+		GetTransferChannelCompleteActionGen().AsAny(),
```

Comment:
How important and neccessary is it that every action is included in this fuzz test? We'll surely add new actions as time goes on, and it may be easy for devs to forget to add a field here for a newly defined action.

Idea (only if it's important that every action is included in this test): define a top-level action set interface, where that interface has a field for every action type. This test then iterates over the fields in said interface to call the appropriate `Get___ActionGen` funcs.

At runtime, check that all actions used in the step files correspond to a field in the interface. If the field is missing, panic.

Just an idea here. Pro is that devs would have a safegaurd against forgetting to add neccessary boilerplate for a new e2e action. Con is increased complexity

Reply:
Neat idea, but I think the added complexity is not worth it. If an action is missing from this test, it's not too bad - if things break, you would notice if you tried to use a test trace with it, and it can be added here then.